# Gene Track绘图函数 (Figure 5E, 6F)
geneSegTrackPlot <- function(DataList, Motif, p.cutoff = 5e-05, TargetGene = "AT1G76550", DataType = c(rep("DAP-seq", 3), rep("RNA-seq", 4)), DAP_yMax = NULL, RNA_yMax = NULL, DAP_fill = "#438AC9", DAP_color = "#7F8283", RNA_fill = "#EF9393", RNA_color = "#7F8283", DataName = c("MYB2", "MYB57", "MYB79", "Control", "MYB2", "MYB57", "MYB79"), pdfFile = "/wrk/chenhao/work/MYB2_2023/RNA_MYB57_79/AT1G76550.pdf", pdf_width = 4, pdf_heigth = 1.8, genome = "/wrk/data/genome/yz_genome_data/aragenome/ara.fa", gffFile = "/wrk/data/genome/yz_genome_data/aragenome/Athaliana.gff3") {
  ## 基因注释信息
  GFF3 <- rtracklayer::import(gffFile) %>% as.data.frame()
  Genome <- readDNAStringSet(genome)
  
  ## 提取目标基因位置等信息
  gene_target_anno <- GFF3[str_detect(GFF3$ID, TargetGene),]
  GR <- gene_target_anno %>% filter(type == "gene") %>% makeGRangesFromDataFrame(keep.extra.columns = F, ignore.strand = T, seqinfo = NULL, seqnames.field = "seqnames", start.field = "start", end.field = "end")
  arrow_df <- data.frame(start = seq(1, width(GR), floor(width(GR) / 5))) %>% mutate(end = start + 59)
  arrow_df <- arrow_df[-c(1, nrow(arrow_df)),]
  arrow_df$start <- arrow_df$start + gene_target_anno$start[gene_target_anno$type == "gene"]
  arrow_df$end <- arrow_df$end + gene_target_anno$start[gene_target_anno$type == "gene"]
  start(GR) <- start(GR) - 1000; width(GR) <- width(GR) + 1000
  
  ## 寻找motif匹配位置
  Seq <- getSeq(Genome, GR)
  motif_match_df <- motifmatchr::matchMotifs(Motif, Seq, out = "positions", p.cutoff = p.cutoff, bg = "even") %>% as.data.frame()
  motif_match_df$start <- motif_match_df$start + start(GR) - 1
  motif_match_df$end <- motif_match_df$end + start(GR) - 1
  
  ## 绘制gene结构图
  geneSegPlot <- ggplot() + 
    # 绘制gene主体结构结构
    geom_rect(data = gene_target_anno[gene_target_anno$type == "gene",], aes(xmin = start, xmax = end, ymin = 1.5, ymax = 4.5), colour = NA, fill = "black") + 
    # 绘制CDS区域
    geom_rect(data = gene_target_anno[gene_target_anno$type == "CDS",], aes(xmin = start, xmax = end, ymin = 0, ymax = 6), colour = NA, fill = "black") + 
    # 绘制motif匹配区域
    geom_rect(data = motif_match_df, aes(xmin = start, xmax = end, ymin = 0, ymax = 6), colour = NA, fill = "red") + 
    # 设定x轴起始终止点
    coord_cartesian(xlim = c(start(GR), end(GR))) + 
    # 绘制箭头以表示基因起始方向
    geom_segment(data = arrow_df, aes(x = start, y = 1.7, xend = end, yend = 3), colour = "white", size = .4) + 
    geom_segment(data = arrow_df, aes(x = end, y = 3, xend = start, yend = 4.3), colour = "white", size = .4) + 
    # 布局
    theme(axis.text = element_blank(), axis.line = element_blank(), axis.ticks = element_blank(), axis.title.y = element_blank(), plot.margin = margin(t = 0, r = 0, b = 0, l = 0)) + 
    labs(x = TargetGene)
  
  ## 读取track coverage
  DfList <- map(DataList, function(Data) {
    Df <- data.frame(Cov = unlist(Data[GR]), Pos = seq_along(unlist(Data[GR])))
  } ) %>% `names<-`(DataName)
  
  ## 根据track coverage设定y轴最大值
  if (is.null(DAP_yMax)) DAP_yMax <- map(DfList[DataType == "DAP-seq"], ~.x$Cov) %>% `names<-`(NULL) %>% do.call(c, .) %>% max() %>% ceiling()
  if (is.null(RNA_yMax)) RNA_yMax <- map(DfList[DataType == "RNA-seq"], ~.x$Cov) %>% `names<-`(NULL) %>% do.call(c, .) %>% max() %>% ceiling()
  
  ## 绘制各样本track
  plotList <- map2(DfList, DataType, function(Df, Type) {
    if (Type == "DAP-seq") color <- DAP_color; fill <- DAP_fill; yMax <- DAP_yMax
    if (Type == "RNA-seq") color <- RNA_color; fill <- RNA_fill; yMax <- RNA_yMax
    
    plot <- ggplot() + geom_area(data = Df, aes(Pos, Cov), fill = fill) + # 绘制track主体结构
      # 绘制track上下边缘，颜色和geom_area保持一致
      geom_line(data = Df, aes(Pos, Cov), color = fill, size = .1) + 
      geom_line(data = Df, aes(Pos, 0), color = color, size = .1) + 
      gg_theme_Publication() + coord_cartesian(ylim = c(0, yMax)) + # 设定y轴最高值
      theme(axis.title.x = element_blank(), axis.title.y = element_blank(), axis.text = element_blank(), axis.line = element_blank(), axis.ticks = element_blank(), plot.margin = margin(t = 0, r = 0, b = 0, l = 0)) + 
      # y轴范围显示
      annotate(geom = "text", x = floor(max(Df$Pos) * .9), y = yMax * .8, size = 3, label = paste0("[0 - ", yMax, "]"), color = "#000000")
  } )
  
  ## 绘制组图
  cmd <- map(seq_along(plotList), ~glue::glue("plotList[[", .x, "]]")) %>% paste(collapse = ", ") %>% paste0("gridExtra::grid.arrange(", ., ", geneSegPlot, ncol = 1, heights = c(rep(1, length(DataList)), .9))")
  
  ## 保存为pdf
  pdf(pdfFile, pdf_width, pdf_heigth)
  eval(parse(text = cmd))
  dev.off()
}
